winch/codegen/src/isa/x64/mod.rs

```diff
@@ -120,15 +120,33 @@ impl TargetIsa for X64 {
         16
     }
 
-    fn host_to_wasm_trampoline(&self, ty: &FuncType) -> Result<MachBufferFinalized<Final>> {
+    fn compile_trampoline(
+        &self,
+        ty: &FuncType,
+        func_index: u32,
+        func_env: &dyn FuncEnv,
+        kind: TrampolineKind,
+    ) -> Result<MachBufferFinalized<Final>> {
+        use TrampolineKind::*;
+
         let abi = abi::X64ABI::default();
         let mut masm = X64Masm::new(self.shared_flags.clone(), self.isa_flags.clone());
         let call_conv = self.wasmtime_call_conv();
 
-        let mut trampoline =
-            Trampoline::new(&mut masm, &abi, regs::scratch(), regs::argv(), &call_conv);
-
-        trampoline.emit_host_to_wasm(ty);
+        let mut trampoline = Trampoline::new(
+            &mut masm,
+            &abi,
+            func_env,
+            regs::scratch(),
+            regs::argv(),
+            &call_conv,
+        );
+
+        match kind {
+            ArrayToWasm => trampoline.emit_array_to_wasm(ty, func_index)?,
+            NativeToWasm => trampoline.emit_native_to_wasm(ty.clone(), func_index)?,
+            WasmToNative => trampoline.emit_wasm_to_native(&ty)?,
```

Comment:
minor nit, but here instead of `ty` vs `ty.clone()` vs `&ty` could these all take `ty: &FuncType`?